src/scanner.c

```diff
@@ -0,0 +1,403 @@
+
+#include <tree_sitter/parser.h>
+#include <ctype.h> // isspace()
+#include <string.h> // memcpy()
+
+enum TokenType {
+  NEWLINE,
+  SEMICOLON,
+  ELSE,
+  RAW_STRING_LITERAL,
+  OPEN_PAREN,
+  CLOSE_PAREN,
+  OPEN_BRACE,
+  CLOSE_BRACE,
+  OPEN_BRACKET,
+  CLOSE_BRACKET,
+  OPEN_BRACKET2,
+  CLOSE_BRACKET2
+};
+
+// ---------------------------------------------------------------------------------------
+// Temporary Stack structure until we can use the `<tree_sitter/array.h>` header from
+// tree sitter 1.0.0. Inspired from tree-sitter-julia.
+
+typedef char Scope;
+
+const Scope SCOPE_TOP_LEVEL = 0;
+const Scope SCOPE_BRACE = 1;
+const Scope SCOPE_PAREN = 2;
+const Scope SCOPE_BRACKET = 3;
+const Scope SCOPE_BRACKET2 = 4;
+
+// A `Stack` data structure for tracking the current `Scope`
+//
+// `SCOPE_TOP_LEVEL` is never actually pushed onto the stack. It is returned from
+// `stack_peek()` as a base case when `len = 0`. Note that in `stack_pop()` we still check
+// for `len > 0` before peeking to retain the invariant that we can't pop without
+// something on the stack.
+// 
+// This actually makes serialization/deserialization very simple. Even if we pushed an
+// initial `SCOPE_TOP_LEVEL` in the create hook, there is no guarantee that that will get
+// serialized (so the length of the stack won't be remembered) because the serialize hook
+// only runs when we accept a token from our external scanner. That would complicate the
+// deserialize hook by forcing us to differentiate between the cases of:
+// 1) A deserialization call restoring state from a previous serialization (len > 0).
+// 2) A deserialization call when there wasn't a previous serialization (len = 0), where
+//    we'd have to repush an initial `SCOPE_TOP_LEVEL`.
+typedef struct {
+  Scope *arr;
+  unsigned len;
+} Stack;
+
+static Stack *new_stack() {
+  Scope *arr = malloc(TREE_SITTER_SERIALIZATION_BUFFER_SIZE);
+  if (arr == NULL) exit(1);
+  Stack *stack = malloc(sizeof(Stack));
+  if (stack == NULL) exit(1);
+  stack->arr = arr;
+  stack->len = 0;
+  return stack;
+}
+
+static void stack_free(Stack *stack) {
+  free(stack->arr);
+  free(stack);
+}
+
+static void stack_push(Stack *stack, Scope scope) {
+  if (stack->len >= TREE_SITTER_SERIALIZATION_BUFFER_SIZE) exit(1);
+  stack->arr[stack->len] = scope;
+  stack->len++;
+}
+
+static Scope stack_peek(Stack *stack) {
+  if (stack->len == 0) {
+    return SCOPE_TOP_LEVEL;
+  } else {
+    return stack->arr[stack->len - 1];
+  }
+}
+
+static void stack_pop(Stack *stack, Scope scope) {
+  if (stack->len == 0) exit(1);
+  Scope x = stack_peek(stack);
+  stack->len--;
+  if (x != scope) exit(1);
```

Comment:
Is this correct? What happens if you write code like `{)`?

Reply:
I believe it is okay. A external scanner token of `)` is not valid (in terms of `valid_symbols`) at that time so you'll never hit the `stack_pop(stack, SCOPE_PAREN)` case.

Instead the `)` falls through and gets matched by the grammar itself, where we catch unmatched closing parenthesis with the rule `")": $ => /\)/,`

I changed the `)` name of the rule to `unmatched_parenthesis` below, otherwise its quite confusing to look at

```
(program [0, 0] - [0, 2]
  ({ [0, 0] - [0, 2]
    body: (unmatched_parenthesis [0, 1] - [0, 2])))
```

Reply:
Ah, ok! Make sense.